File: unsupported/juno/lit/juno/reduce_conditional.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// RUN: %juno --gen-js -O %s | %FileCheck %s --match-full-lines

function foo(x, y) {
  return true ? x + 1 : y + 1
}

// CHECK-LABEL: function foo(x, y) {
// CHECK-NEXT:   return x + 1;
// CHECK-NEXT: }

function bar(x, y) {
  return false ? x + 1 : y + 1
}

// CHECK-LABEL: function bar(x, y) {
// CHECK-NEXT:   return y + 1;
// CHECK-NEXT: }

function if_false(x, y) {
  if (false) {
    return x;
  }
  return y;
}

// CHECK-LABEL: function if_false(x, y) {
// CHECK-NEXT:   return y;
// CHECK-NEXT: }

function if_true(x, y) {
  if (true) {
    return x;
  }
  return y;
}

// CHECK-LABEL: function if_true(x, y) {
// CHECK-NEXT:   {
// CHECK-NEXT:     return x;
// CHECK-NEXT:   }
// CHECK-NEXT:   return y;
// CHECK-NEXT: }

function if_false_else(x, y) {
  if (false) {
    return x;
  } else {
    return y;
  }
}

// CHECK-LABEL: function if_false_else(x, y) {
// CHECK-NEXT:   {
// CHECK-NEXT:     return y;
// CHECK-NEXT:   }
// CHECK-NEXT: }

function if_true_else(x, y) {
  if (true) {
    return x;
  } else {
    return y;
  }
}

// CHECK-LABEL: function if_true_else(x, y) {
// CHECK-NEXT:   {
// CHECK-NEXT:     return x;
// CHECK-NEXT:   }
// CHECK-NEXT: }

function for_if_false(x, y) {
  for (;;) if (false) {}
}

// CHECK-LABEL: function for_if_false(x, y) {
// CHECK-NEXT:   for(;;)
// CHECK-NEXT:     ;
// CHECK-NEXT: }

File: unsupported/juno/crates/juno_pass/src/passes/reduce_conditional.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Pass for optimizing conditional expressions where the conditional is a literal.
//!
//! Operates on both conditional expressions and `if` statements (with or without `else`).
//! For example, transforms
//! ```js
//! true ? a : b
//! ```
//! into
//! ```js
//! a
//! ```

use juno::ast::*;

use crate::Pass;

#[derive(Default)]
pub struct ReduceConditional {}

impl ReduceConditional {
    pub fn new() -> Self {
        Default::default()
    }
}

impl Pass for ReduceConditional {
    fn name(&self) -> &'static str {
        "Reduce conditional"
    }
    fn description(&self) -> &'static str {
        "Transforms literal conditionals into whichever branch is taken"
    }
    fn run<'gc>(
        &mut self,
        gc: &'gc GCLock<'_, '_>,
        node: &'gc Node<'gc>,
    ) -> TransformResult<&'gc Node<'gc>> {
        VisitorMut::call(self, gc, node, None)
    }
}

impl<'gc> VisitorMut<'gc> for ReduceConditional {
    fn call(
        &mut self,
        lock: &'gc GCLock,
        node: &'gc Node<'gc>,
        _parent: Option<Path<'gc>>,
    ) -> TransformResult<&'gc Node<'gc>> {
        match node {
            Node::ConditionalExpression(ConditionalExpression {
                test: Node::BooleanLiteral(BooleanLiteral { value, .. }),
                consequent,
                alternate,
                ..
            })
            | Node::IfStatement(IfStatement {
                test: Node::BooleanLiteral(BooleanLiteral { value, .. }),
                consequent,
                alternate: Some(alternate),
                ..
            }) => {
                node.replace_with_existing(if *value { consequent } else { alternate }, lock, self)
            }
            Node::IfStatement(IfStatement {
                test: Node::BooleanLiteral(BooleanLiteral { value, .. }),
                consequent,
                alternate: None,
                ..
            }) => {
                if *value {
                    node.replace_with_existing(consequent, lock, self)
                } else {
                    TransformResult::Removed
                }
            }
            _ => node.visit_children_mut(lock, self),
        }
    }
}
